float64 range
float64 range_max
float64 angle
float64 angle_max

